main.go

```diff
@@ -16,33 +16,269 @@ limitations under the License.
 
 package main
 
-import "github.com/googleinterns/k8s-topology-simulator/modeling"
+import (
+	"encoding/csv"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/googleinterns/k8s-topology-simulator/modeling"
+)
 
 func main() {
-	zoneA := modeling.Zone{Nodes: 30, Endpoints: 60, Name: "a"}
-	zoneB := modeling.Zone{Nodes: 35, Endpoints: 70, Name: "b"}
-	zoneC := modeling.Zone{Nodes: 50, Endpoints: 80, Name: "c"}
-	zones := []modeling.Zone{zoneA, zoneB, zoneC}
+	// algorithm name, default shared global
+	algPtr := flag.String("alg", "SharedGlobalAlgorithm", "routing algorithm")
+	// input file
+	inputPtr := flag.String("input", "test.csv", "info of each zone")
+	// output file, default alg_result.csv
+	outputPtr := flag.String("output", "", "simulation results")
+	flag.Parse()
+	if *outputPtr == "" {
+		*outputPtr = *algPtr + "_result.csv"
+	}
 
-	alg, err := modeling.CreateAlg(0.4, 100)
+	var errList []<-chan error
+
+	zoneNames, inputQueue, errC, err := parseInput(*inputPtr)
 	if err != nil {
 		panic(err)
 	}
-
-	var sim modeling.TheoreticalSimulator
-
-	// TODO: operations on metrics rather than printing only
-	model, err := modeling.NewModel(zones, alg, sim)
+	errList = append(errList, errC)
+	outputQueue, errC, err := startSimulation(*algPtr, inputQueue)
 	if err != nil {
 		panic(err)
 	}
-
-	err = model.StartSimulation()
+	errList = append(errList, errC)
+	errC, err = parseResult(*outputPtr, zoneNames, outputQueue)
 	if err != nil {
 		panic(err)
 	}
-	err = model.PrintLastResult()
+	errList = append(errList, errC)
+	// wait for goroutings to return
+	err = waitForSimulation(errList...)
 	if err != nil {
 		panic(err)
 	}
 }
+
+// every row of the input file will be parsed to one instance of inputData
+type inputData struct {
+	// input id of the row
+	id string
+	// parse zone info of the input file into zone data structure
+	zones []modeling.Zone
+}
+
+// every piece of input data will be processed to one instance of outputData
```

Comment:
Nit: maybe a couple tiny changes:
```suggestion
// every instance of inputData will be mapped to one instance of outputData
```

Reply:
changed